### PR TITLE
Fix https://github.com/wso2/product-apim/issues/3883

### PR DESCRIPTION
**Pull request by kavishkafernando**
## Documentation
A new tag name isDefault has been introduced for the Environments configuration in api-manager.xml.
E.g : <Environment type="production" isDefault=“true”>
If multiple production or sandbox environments are introduced the isDefault= true tag can be used to define the default sandbox/ production environment.
The default production/ sandbox environments will then be shown in the relavant tab.

     <Environment type="production" isDefault=“true”>
           <Name>Production</Name>
           <ServerURL>https://localhost:${mgt.transport.https.port}/services/</ServerURL>
           <Username>admin</Username>
           <Password>admin</Password>
           <GatewayEndpoint>http://production234:${http.nio.port},https://production234:${https.nio.port}</GatewayEndpoint>
        </Environment>  
            <Environment type="production">
           <Name>Production1</Name>
           <ServerURL>https://localhost:${mgt.transport.https.port}/services/</ServerURL>
           <Username>admin</Username>
           <Password>admin</Password>
           <GatewayEndpoint>http://production:${http.nio.port},https://production:${https.nio.port}</GatewayEndpoint>
        </Environment>  
        <Environment type="sandbox" isDefault=“true>
           <Name>Sandbox</Name>
           <ServerURL>https://localhost:${mgt.transport.https.port}/services/</ServerURL>
           <Username>admin</Username>
           <Password>admin</Password>        
           <GatewayEndpoint>http://sandbox234:${http.nio.port},https://sandbox234:${https.nio.port}</GatewayEndpoint>
       </Environment>
            <Environment type="sandbox">
           <Name>Sandbox1</Name>
           <ServerURL>https://localhost:${mgt.transport.https.port}/services/</ServerURL>
           <Username>admin</Username>
           <Password>admin</Password>        
           <GatewayEndpoint>http://sandbox:${http.nio.port},https://sandbox:${https.nio.port}</GatewayEndpoint>
       </Environment>

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 1]
```diff
@@ -2147,6 +2147,64 @@ public static NativeArray jsFunction_getHTTPsGatewayEndpointURLs(Context cx, Scr
...
+        for (Environment environment : environments.values()) {
+            if (APIConstants.GATEWAY_ENV_TYPE_HYBRID.equals(environment.getType())) {
+                json.put(APIConstants.GATEWAY_ENV_TYPE_PRODUCTION,
+                        APIStoreHostObject.getHttpsEnvironmentUrl(environment));
+                json.put(APIConstants.GATEWAY_ENV_TYPE_SANDBOX,
+                        APIStoreHostObject.getHttpsEnvironmentUrl(environment));
+                return json;
+            } else {
+                String environmentType = APIConstants.GATEWAY_ENV_TYPE_PRODUCTION.equals(environment.getType())
+                        ? "production" : "sandbox";
```
Comment (tgtshanika) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Can't we use the same constant values (GATEWAY_ENV_TYPE_PRODUCTION, GATEWAY_ENV_TYPE_SANDBOX) here as well

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 2]
```diff
@@ -2147,6 +2147,64 @@ public static NativeArray jsFunction_getHTTPsGatewayEndpointURLs(Context cx, Scr
...
+                if (environment.isDefault()) {
+                    json.put(environmentType,
+                            APIStoreHostObject.getHttpsEnvironmentUrl(environment));
+                }
+            }
+        }
+
+        if (json.get("production") == null) {
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Shall we move "production" to a String constant?

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 3]
```diff
@@ -2147,6 +2147,64 @@ public static NativeArray jsFunction_getHTTPsGatewayEndpointURLs(Context cx, Scr
...
+            for (Environment environment : environments.values()) {
+                if (APIConstants.GATEWAY_ENV_TYPE_PRODUCTION.equals(environment.getType())) {
+                    json.put(APIConstants.GATEWAY_ENV_TYPE_PRODUCTION,
+                            APIStoreHostObject.getHttpsEnvironmentUrl(environment));
+                    break;
+                }
+            }
+        }
+
+        if (json.get("sandbox") == null) {
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Lets move "sandbox" to a String constant.

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 4]
```diff
@@ -2147,6 +2147,64 @@ public static NativeArray jsFunction_getHTTPsGatewayEndpointURLs(Context cx, Scr
...
+                }
+            }
+        }
+
+        return json;
+    }
+
+    private static String getHttpsEnvironmentUrl(Environment environment) {
+        for (String url : environment.getApiGatewayEndpoint().split(",")) {
+            if (url.startsWith("https:")) {
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Lets move "https:" to a String constant.

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 5]
```diff
@@ -2147,6 +2147,64 @@ public static NativeArray jsFunction_getHTTPsGatewayEndpointURLs(Context cx, Scr
         return myn;
     }
 
+    public static JSONObject jsFunction_getHTTPSGatewayEndpointURLsWithType(Context cx, Scriptable thisObj,
+                                                                            Object[] args, Function funObj)
+            throws ScriptException, APIManagementException {
+
+        APIManagerConfiguration config = HostObjectComponent.getAPIManagerConfiguration();
+        Map<String, Environment> environments = config.getApiGatewayEnvironments();
+        JSONObject json = new JSONObject();
```
Comment (tgtshanika) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Shall we use a meaningful name?

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java (hunk 1 above)
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: Lets use String constants here.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java [hunk 6]
```diff
@@ -200,6 +200,12 @@ private void readChildElements(OMElement serverConfig,
                     } else {
                         environment.setShowInConsole(true);
                     }
+                    String isDefault = environmentElem.getAttributeValue(new QName("isDefault"));
+                    if (isDefault != null) {
```
Comment (tgtshanika) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java]: Shall we use StringUtils.isEmpty() ?

Comment (isharac) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java]: here we only needed to check null.

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 7]
```diff
@@ -2147,6 +2147,64 @@ public static NativeArray jsFunction_getHTTPsGatewayEndpointURLs(Context cx, Scr
...
+        if (json.get("sandbox") == null) {
+            for (Environment environment : environments.values()) {
+                if (APIConstants.GATEWAY_ENV_TYPE_SANDBOX.equals(environment.getType())) {
+                    json.put(APIConstants.GATEWAY_ENV_TYPE_SANDBOX,
+                            APIStoreHostObject.getHttpsEnvironmentUrl(environment));
+                    break;
+                }
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java [hunk 8]
```diff
@@ -2147,6 +2147,64 @@ public static NativeArray jsFunction_getHTTPsGatewayEndpointURLs(Context cx, Scr
...
+    private static String getHttpsEnvironmentUrl(Environment environment) {
+        for (String url : environment.getApiGatewayEndpoint().split(",")) {
+            if (url.startsWith("https:")) {
+                return url;
+            }
+        }
+        return "";
+    }
+
+
```
[the remark this answers is not in the document]
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java (hunk 1 above)
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java (hunk 5 above)
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.hostobjects/src/main/java/org/wso2/carbon/apimgt/hostobjects/APIStoreHostObject.java]: fixed

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java (hunk 6 above)
Comment (kavishkafernando) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java]: fixed

**Pull request merged**